network/framework/src/application/storage.rs

```diff
@@ -91,6 +103,69 @@ impl PeersAndMetadata {
         all_peers
     }
 
+    /// Return copy of list of peers if generation is different than previously held.
+    /// May optionally filter for connected peers only.
+    /// May optionally filter on match for any of a set of ProtocolId.
+    pub fn get_all_peers_generational(
+        &self,
+        generation: u32,
+        require_connected: bool,
+        protocol_ids: &[ProtocolId],
+    ) -> Option<(Vec<PeerNetworkId>, u32)> {
+        let generation_test = self.generation.load(Ordering::SeqCst);
+        if generation == generation_test {
+            return None;
+        }
+        let mut all_peers = Vec::new();
+        let read = self.peers_and_metadata.read();
```

Comment:
Silly question: if we move to the generational API, we should remove `cached_peers_and_metadata` entirely, right? Why keep them both around if they're providing the same functionality?